contracts/vortex/CarbonVortex.sol

```diff
@@ -105,31 +172,171 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
+     */
+    function setMaxTargetTokenSaleAmount(
+        uint128 newMaxTargetTokenSaleAmount
+    ) external onlyAdmin greaterThanZero(newMaxTargetTokenSaleAmount) {
+        _setMaxTargetTokenSaleAmount(newMaxTargetTokenSaleAmount);
+    }
+
+    /**
+     * @notice sets the min target token sale amount
+     *
+     * requirements:
+     *
+     * - the caller must be the admin of the contract
+     */
+    function setMinTargetTokenSaleAmount(
+        uint128 newMinTargetTokenSaleAmount
+    ) external onlyAdmin greaterThanZero(newMinTargetTokenSaleAmount) {
+        _setMinTokenSaleAmount(_targetToken, newMinTargetTokenSaleAmount);
+    }
+
+    /**
+     * @notice sets if trading is enabled or disabled for a token
+     *
+     * requirements:
+     *
+     * - the caller must be the admin of the contract
+     */
+    function disablePair(Token token) external onlyAdmin {
+        _setPairDisabled(token);
+    }
+
+    /**
+     * @dev withdraws funds held by the contract and sends them to an account
+     *
+     * requirements:
+     *
+     * - the caller must be the admin of the contract
+     */
+    function withdrawFunds(
+        Token[] calldata tokens,
+        address payable target,
+        uint256[] calldata amounts
+    ) external validAddress(target) validateTokens(tokens) nonReentrant onlyAdmin {
+        uint256 len = tokens.length;
+        if (len != amounts.length) {
+            revert InvalidAmountLength();
+        }
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            // safe due to nonReentrant modifier (forwards all available gas in case of ETH)
+            tokens[i].unsafeTransfer(target, amounts[i]);
+        }
+
+        emit FundsWithdrawn({ tokens: tokens, caller: msg.sender, target: target, amounts: amounts });
+    }
+
     /**
      * @inheritdoc ICarbonVortex
      */
-    function rewardsPPM() external view returns (uint256) {
+    function rewardsPPM() external view returns (uint32) {
         return _rewardsPPM;
     }
 
     /**
      * @inheritdoc ICarbonVortex
      */
-    function totalBurned() external view returns (uint256) {
-        return _totalBurned;
+    function totalCollected() external view returns (uint256) {
+        return _totalCollected;
     }
 
     /**
      * @inheritdoc ICarbonVortex
      */
     function availableTokens(Token token) external view returns (uint256) {
-        return _carbonController.accumulatedFees(token) + token.balanceOf(address(this));
+        uint256 totalFees = 0;
+        if (address(_carbonController) != address(0)) {
+            totalFees += _carbonController.accumulatedFees(token);
+        }
+        if (address(_vault) != address(0)) {
+            totalFees += token.balanceOf(address(_vault));
+        }
+        if (address(_oldVortex) != address(0)) {
+            totalFees += token.balanceOf(address(_oldVortex));
+        }
+        return totalFees + token.balanceOf(address(this));
```

Comment:
Note that `token` can be a rouge token, and this `balanceOf` can be used for a read-only reentrancy (i.e., attacking another contract). No need to address it, IMHO.